pkg/secret/ensure.go

```
@@ -46,3 +60,85 @@ func Ensure(client kubernetes.Interface, namespace string, secret *v1.Secret) (*
 }
+
+func EnsureFromSA(client kubernetes.Interface, saName, namespace string) (*v1.Secret, error) {
+	sa, err := client.CoreV1().ServiceAccounts(namespace).Get(context.TODO(), saName, metav1.GetOptions{})
+	if err != nil {
+		return nil, errors.Wrapf(err, "failed to get ServiceAccount %s/%s", namespace, saName)
+	}
+
+	saSecret, err := getSecretFromSA(client, sa)
+
+	if err == nil {
+		return saSecret,nil
+	}
+
+	// We couldn't find right secret from this SA, search all Secrets
+	saSecret, err = getSecretForSA(client, sa)
+	if err != nil {
+		newSecret := &v1.Secret{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:      fmt.Sprintf("%s-token-%s", sa.Name, generateRandomString(5)),
+				Namespace: namespace,
+				Annotations: map[string]string{
+					saAnnotation:        saName,
+					createdByAnnotation: creatorName,
+				},
+			},
+			Type: v1.SecretTypeServiceAccountToken,
+		}
+
+		saSecret, err = Ensure(client, newSecret.Namespace, newSecret)
+	}
+
+	secretRef := v1.ObjectReference{
+		Name: saSecret.Name,
+	}
+
+	sa.Secrets = append(sa.Secrets, secretRef)
+	_, err = serviceaccount.Ensure(client, namespace, sa)
+	if err != nil {
+		return nil, errors.Wrapf(err, "failed to update ServiceAccount %v with Secret reference %v", saName, secretRef.Name)
+	}
+
+	return saSecret, err
+}
+
+func getSecretFromSA(client kubernetes.Interface, sa *v1.ServiceAccount) (*v1.Secret, error) {
+	secretNamePrefix := fmt.Sprintf("%s-token-", sa.Name)
+	for _, saSecretRef := range sa.Secrets {
+		if strings.HasPrefix(saSecretRef.Name, secretNamePrefix) {
+			saSecret, _ := client.CoreV1().Secrets(sa.Namespace).Get(context.TODO(), saSecretRef.Name, metav1.GetOptions{})
+			if saSecret.Annotations[saAnnotation] == sa.Name && saSecret.Type == v1.SecretTypeServiceAccountToken {
+				return saSecret, nil
+			}
+		}
+	}
+
+	return nil, errors.New(fmt.Sprintf("secret token for service-account %v/%v not found", sa.Namespace, sa.Name))
```

We don't need to return an error if not found - simply return nil `*v1.Secret` . There's no actual error that can occur so remove it as a return value.